File: code/packages/fsharp/wasm-simulator/tests/CodingAdventures.WasmSimulator.Tests/WasmSimulatorTests.fs
module CodingAdventures.WasmSimulator.Tests

open CodingAdventures.WasmSimulator.FSharp
open Xunit

type WasmSimulatorTests() =
    [<Fact>]
    member _.``Has version``() =
        Assert.Equal("0.1.0", Version.VERSION)

    [<Fact>]
    member _.``Encoding helpers produce expected bytes``() =
        Assert.Equal<byte>([| 0x6Auy |], encodeI32Add ())
        Assert.Equal<byte>([| 0x21uy; 0x02uy |], encodeLocalSet 2)

    [<Fact>]
    member _.``Decoder reads i32 const``() =
        let decoder = WasmDecoder()
        let instruction = decoder.Decode(encodeI32Const 42, 0)
        Assert.Equal("i32.const", instruction.Mnemonic)
        Assert.Equal(42, instruction.Operand.Value)
        Assert.Equal(5, instruction.Size)

    [<Fact>]
    member _.``Simulator runs simple program``() =
        let simulator = WasmSimulator(4)
        let program =
            assembleWasm
                [
                    encodeI32Const 1
                    encodeI32Const 2
                    encodeI32Add ()
                    encodeLocalSet 0
                    encodeEnd ()
                ]

        let traces = simulator.Run(program)
        Assert.Equal(5, traces.Length)
        Assert.Equal(3, simulator.Locals[0])
        Assert.True(simulator.Halted)

File: code/packages/csharp/wasm-simulator/tests/CodingAdventures.WasmSimulator.Tests/WasmSimulatorTests.cs
using System;
using CodingAdventures.WasmSimulator;
using Xunit;

namespace CodingAdventures.WasmSimulator.Tests;

public class WasmSimulatorTests
{
    [Fact]
    public void HasVersion() => Assert.Equal("0.1.0", WasmSimulatorVersion.VERSION);

    [Fact]
    public void EncodingHelpersProduceExpectedBytes()
    {
        Assert.Equal(new byte[] { 0x6A }, WasmSimulator.EncodeI32Add());
        Assert.Equal(new byte[] { 0x21, 0x02 }, WasmSimulator.EncodeLocalSet(2));
    }

    [Fact]
    public void DecoderReadsI32Const()
    {
        var decoder = new WasmDecoder();
        var instruction = decoder.Decode(WasmSimulator.EncodeI32Const(42), 0);
        Assert.Equal("i32.const", instruction.Mnemonic);
        Assert.Equal(42, instruction.Operand);
        Assert.Equal(5, instruction.Size);
    }

    [Fact]
    public void SimulatorRunsSimpleProgram()
    {
        var simulator = new WasmSimulator(4);
        var program = WasmSimulator.AssembleWasm(
        [
            WasmSimulator.EncodeI32Const(1),
            WasmSimulator.EncodeI32Const(2),
            WasmSimulator.EncodeI32Add(),
            WasmSimulator.EncodeLocalSet(0),
            WasmSimulator.EncodeEnd(),
        ]);

        var traces = simulator.Run(program);
        Assert.Equal(5, traces.Count);
        Assert.Equal(3, simulator.Locals[0]);
        Assert.True(simulator.Halted);
    }
}
